cpp/src/arrow/flight/transport/grpc/grpc_client.cc

```diff
@@ -0,0 +1,872 @@
+    default_args[GRPC_ARG_USE_LOCAL_SUBCHANNEL_POOL] = 1;
+
+    if (options.override_hostname != "") {
+      args.SetSslTargetNameOverride(options.override_hostname);
+    }
+
+    // Allow setting generic gRPC options.
+    for (const auto& arg : options.generic_options) {
+      if (util::holds_alternative<int>(arg.second)) {
+        default_args[arg.first] = util::get<int>(arg.second);
+      } else if (util::holds_alternative<std::string>(arg.second)) {
+        args.SetString(arg.first, util::get<std::string>(arg.second));
+      }
+      // Otherwise unimplemented
+    }
+    for (const auto& pair : default_args) {
+      args.SetInt(pair.first, pair.second);
+    }
+
+    std::vector<std::unique_ptr<::grpc::experimental::ClientInterceptorFactoryInterface>>
+        interceptors;
+    interceptors.emplace_back(
+        new GrpcClientInterceptorAdapterFactory(std::move(options.middleware)));
+
+    stub_ = pb::FlightService::NewStub(
+        ::grpc::experimental::CreateCustomChannelWithInterceptors(
+            grpc_uri.str(), creds, args, std::move(interceptors)));
+    return Status::OK();
+  }
+
+  Status Close() override {
+    // TODO(ARROW-15473): if we track ongoing RPCs, we can cancel them first
+    // gRPC does not offer a real Close(). We could reset() the gRPC
+    // client but that can cause gRPC to hang in shutdown
+    // (ARROW-15793).
+    return Status::OK();
+  }
+
+  Status Authenticate(const FlightCallOptions& options,
+                      std::unique_ptr<ClientAuthHandler> auth_handler) override {
+    auth_handler_ = std::move(auth_handler);
+    ClientRpc rpc(options);
+    std::shared_ptr<
+        ::grpc::ClientReaderWriter<pb::HandshakeRequest, pb::HandshakeResponse>>
+        stream = stub_->Handshake(&rpc.context);
+    GrpcClientAuthSender outgoing{stream};
+    GrpcClientAuthReader incoming{stream};
+    RETURN_NOT_OK(auth_handler_->Authenticate(&outgoing, &incoming));
+    // Explicitly close our side of the connection
+    bool finished_writes = stream->WritesDone();
+    RETURN_NOT_OK(internal::FromGrpcStatus(stream->Finish(), &rpc.context));
+    if (!finished_writes) {
+      return MakeFlightError(FlightStatusCode::Internal,
+                             "Could not finish writing before closing");
+    }
+    return Status::OK();
+  }
+
+  arrow::Result<std::pair<std::string, std::string>> AuthenticateBasicToken(
+      const FlightCallOptions& options, const std::string& username,
+      const std::string& password) override {
+    // Add basic auth headers to outgoing headers.
+    ClientRpc rpc(options);
+    internal::AddBasicAuthHeaders(&rpc.context, username, password);
+
+    std::shared_ptr<
+        ::grpc::ClientReaderWriter<pb::HandshakeRequest, pb::HandshakeResponse>>
+        stream = stub_->Handshake(&rpc.context);
+    GrpcClientAuthSender outgoing{stream};
+    GrpcClientAuthReader incoming{stream};
```

Comment:
Nothing seems to be done with these? Also I'm not sure why basic auth isn't implemented as a dedicated `ClientAuthHandler` instead of having this separate method (though this may be not new in this PR).

Reply:
Not new in this PR. I want to consolidate the auth methods at some point since there were several added over the years.

Reply:
Especially as other methods call `rpc.SetToken(auth_handler_.get())` and that would only send the basic auth token if the auth handler is not null, right?

Reply:
I think the way it was intended to be used, this is basically "login" and the returned token is configured and passed back via middleware or an auth handler

Reply:
How would the auth handler get the token, though, if it isn't called here?

Reply:
The user would have to configure it manually. (Not a very convenient solution, agreed.)